x/evm/keeper/state_transition.go

```diff
@@ -446,12 +481,78 @@ func (k *Keeper) ApplyMessageWithConfig(
 	// reset leftoverGas, to be used by the tracer
 	leftoverGas = msg.GasLimit - gasUsed
 
+	var replyLog StateDBGetLogsReply
+	sgxRpcClient.StateDBGetLogs(StateDBGetLogsArgs{}, &replyLog)
+
 	return &types.MsgEthereumTxResponse{
 		GasUsed:   gasUsed,
 		VmError:   vmError,
 		Ret:       ret,
-		Logs:      types.NewLogsFromEth(stateDB.Logs()),
+		Logs:      types.NewLogsFromEth(replyLog.Logs),
 		Hash:      cfg.TxConfig.TxHash.Hex(),
 		BlockHash: ctx.HeaderHash(),
 	}, nil
 }
+
+// prepareTxForSgx prepares the transaction for the SGX enclave. It:
+//   - creates an RPC server around the keeper to receive requests sent by the
+//     SGX
+//   - sends a "PrepareTx" request to the SGX enclave with the relevant tx and
+//     block info
+func (k *Keeper) prepareTxForSgx(ctx sdk.Context, msg core.Message, cfg *EVMConfig, sgxRpcClient *sgxRpcClient) error {
+	// Step 1. Create an RPC server to receive requests from the SGX enclave.
+	err := k.runRPCServer(ctx, msg, cfg)
+	if err != nil {
+		return err
+	}
+
+	// Step 2. Send a "PrepareTx" request to the SGX enclave.
+	chainConfigJson, err := json.Marshal(cfg.ChainConfig)
+	if err != nil {
+		return err
+	}
+	ctx.HeaderHash()
+	args := PrepareTxArgs{
+		Header: ctx.BlockHeader(),
+		Msg:    msg,
+		EvmConfig: PrepareTxEVMConfig{
+			ChainConfigJson: chainConfigJson,
+			CoinBase:        cfg.CoinBase,
+			BaseFee:         cfg.BaseFee,
+			TxConfig:        cfg.TxConfig,
+			DebugTrace:      cfg.DebugTrace,
+			NoBaseFee:       cfg.FeeMarketParams.NoBaseFee,
+			EvmDenom:        cfg.Params.EvmDenom,
+			Overrides:       ToGOB64(cfg.Overrides),
+		},
+	}
+
+	return sgxRpcClient.PrepareTx(args, &PrepareTxReply{})
+}
+
+func (k *Keeper) runRPCServer(ctx sdk.Context, msg core.Message, cfg *EVMConfig) error {
```

Comment:
I know I created this function, but I think it's super ugly to put here.

IMO, we should run the RPC server in `server/start.go`, like all other servers. Then register this RPC server in `RegisterServices` or something like that.